OPHAKTKEH-305

pkoivisto opened the pull request.

pkoivisto committed:
Disable the next button independently from the error label shown on the form

pkoivisto committed:
Display helper text (amount of characters used) for the message also on the preview step page.

p3j4m commented on src/main/reactjs/src/components/contactRequest/steps/WriteMessage.tsx:
```diff
@@ -8,7 +8,7 @@ import {
 } from 'components/contactRequest/ContactRequestFormUtils';
```
Tuossa on nyt sellainen ongelma/ominaisuus, minkä takia error viesti ei näy silloin kun mennään yksi steppi edestakaisin.

p3j4m replied:
yksi vaihtoehto olisi laittaa handleMessageFieldErrors useEffectiin, silloin sitä tarkistetaan myös steppiä vaihtaessa

p3j4m replied:
Kokeilin nopeasti jotain tällaista:

```
import { useEffect, useRef, useState } from 'react';

import {
  ChosenTranslators,
  ChosenTranslatorsHeading,
  DisplayContactInfo,
  StepHeading,
} from 'components/contactRequest/ContactRequestFormUtils';
import { CustomTextField } from 'components/elements/CustomTextField';
import { H3 } from 'components/elements/Text';
import { translateOutsideComponent, useAppTranslation } from 'configs/i18n';
import { useAppDispatch, useAppSelector } from 'configs/redux';
import { TextFieldTypes } from 'enums/app';
import { ContactRequestFormStep } from 'enums/contactRequest';
import { useWindowProperties } from 'hooks/useWindowProperties';
import { setContactRequest } from 'redux/actions/contactRequest';
import { contactRequestSelector } from 'redux/selectors/contactRequest';
import { Utils } from 'utils';
import { StringUtils } from 'utils/string';

const getErrorForMessage = (message: string) => {
  const t = translateOutsideComponent();
  const error = Utils.inspectCustomTextFieldErrors(
    TextFieldTypes.Textarea,
    message,
    true
  );

  return error ? t(`akt.${error}`) : '';
};

export const WriteMessage = ({
  disableNext,
}: {
  disableNext: (disabled: boolean) => void;
}) => {
  // I18n
  const { t } = useAppTranslation({
    keyPrefix: 'akt.component.contactRequestForm',
  });
  const messageFieldRef = useRef<HTMLInputElement>();

  //Windows properties
  const { isPhone } = useWindowProperties();

  // State
  const [fieldError, setFieldError] = useState('');

  // Redux
  const { request } = useAppSelector(contactRequestSelector);
  const dispatch = useAppDispatch();

  useEffect(() => {
    const hasBlankMessage = StringUtils.isBlankString(request?.message);
    // Instead of relying on fieldError in local state, compute the error text
    // separately to ensure that we can disable the next button independently of
    // showing the error labels. The error label is to be shown only
    // after an onBlur event has occurred.
    const hasFieldError = getErrorForMessage(request?.message || '').length > 0;
    handleMessageFieldErrors();
    disableNext(hasBlankMessage || hasFieldError);
  }, [disableNext, fieldError, request]);

  const handleMessageFieldChange = () => {
    if (fieldError) {
      handleMessageFieldErrors();
    }

    dispatch(setContactRequest({ message: messageFieldRef.current?.value }));
  };

  const handleMessageFieldErrors = () => {
    if (messageFieldRef.current?.value)
      setFieldError(getErrorForMessage(messageFieldRef?.current?.value));
  };

  const getHelperMessage = () => {
    const value = request?.message;
    const errorToShow = fieldError ? `${fieldError}.` : '';
    const maxLength = Utils.getMaxTextAreaLength();

    return `${errorToShow} ${value?.length} / ${maxLength} ${t('characters')}`;
  };

  return (
    <div className="rows">
      <StepHeading step={ContactRequestFormStep.WriteMessage} />
      <div className="rows gapped">
        <ChosenTranslatorsHeading />
        <ChosenTranslators />
        {!isPhone && <DisplayContactInfo />}
        <div className="rows gapped">
          <H3>
            {t(
              `steps.${
                ContactRequestFormStep[ContactRequestFormStep.WriteMessage]
              }`
            )}
          </H3>
          <CustomTextField
            inputRef={messageFieldRef}
            id="contact-request-page__message-field"
            label={t('formLabels.writeMessageHere')}
            value={request?.message}
            type={TextFieldTypes.Textarea}
            onChange={handleMessageFieldChange}
            showHelperText
            helperText={getHelperMessage()}
            error={fieldError.length > 0}
            multiline
            fullWidth
            required
          />
        </div>
      </div>
    </div>
  );
};
```

pkoivisto replied:
Joo, tämä ongelma oli kyllä tiedossa. Yksi vaihtoehto voisi tietty olla siirtää kenttien errorit reduxiin, jolloin ei ehkä tarvitsisi kikkailla `useRef` kanssa.

p3j4m replied:
jep, sekin on yksi vaihtoehto, mutta tässä kuitenkin `useRef` toimii aika hyvin?

pkoivisto replied:
Ehkä sen saa toimimaan. Ihan täydellisesti ei vielä tuo versio toiminut. :)
![Screenshot 2022-04-06 at 9 25 28](https://user-images.githubusercontent.com/18038608/161909286-cd51e0f8-b600-4e4d-9b5b-a42548af9fbf.png)

Kokeilen vähän eri vaihtoehtoja.

pkoivisto replied:
Mä muutin tuon nyt niin että virheviesti pidetään reduxissa `messageError` kentässä. Virheen pitäminen lokaalissa tilassa siten että se toimii konsistentisti onBlurin, onChangen ja stepperin kelauksen kanssa tuntui vaan liian monimutkaiselta.

pkoivisto committed:
Store messageError in redux as part of ContactRequestState

pkoivisto committed:
Fix error message test.

For some reason, the pasted text didn't "stick" to the input, but was rather cleared away with onBlur.
Sending an enter keypress to the element forces the issue.

kimmotaskinen commented:
Ei liity suoraan tähän PR:ään mutta contact details email kentän regex voisi päivittää, esim: https://emailregex.com/

Nyt pääsee esim, asdf@asd..com (<- tupla pisteet) läpi

pkoivisto commented:
> Ei liity suoraan tähän PR:ään mutta contact details email kentän regex voisi päivittää, esim: https://emailregex.com/
> 
> Nyt pääsee esim, asdf@asd..com (<- tupla pisteet) läpi

Jep, tänhetkinen versio on todella naiivi tsekkaus. Näitä eri vaihtoehtoja on palloteltu muutamaan otteeseen, ja nykyiseen versioon päädyttiin huomattavasti monimutkaisemmasta muistaakseni mm. perffisyistä: monimutkaisempi regex jumitti selaimen aivan täysin kun validoitavan emailin pituus oli kenties jossain 30-40 merkin tienoilla. Vahvat eksponentiaalisen kasvun vibat tuli siitä.

Sitten tietysti on myös se ongelma, että validaatiolla ei voi varmistua siitä onko mailiosoite oikeasti käytössä / kuuluu kyseiselle henkilölle. Ei toki tarkoita etteikö nykytilaa voisi parantaa. :)

kimmotaskinen commented on src/main/reactjs/src/components/contactRequest/steps/WriteMessage.tsx:
```diff
@@ -8,74 +8,71 @@ import {
+  const t = translateOutsideComponent();
+  const error = Utils.inspectCustomTextFieldErrors(
+    TextFieldTypes.Textarea,
+    message || '',
```
Oisko tuo selkeämpi, että passataan vain message ja utils funkkari sitten katsoo, että onko undefined vai string?

pkoivisto replied:
Joo, eiköhän se ole. Muutan.

kimmotaskinen commented:
> > Ei liity suoraan tähän PR:ään mutta contact details email kentän regex voisi päivittää, esim: https://emailregex.com/
> > Nyt pääsee esim, asdf@asd..com (<- tupla pisteet) läpi
> 
> Jep, tänhetkinen versio on todella naiivi tsekkaus. Näitä eri vaihtoehtoja on palloteltu muutamaan otteeseen, ja nykyiseen versioon päädyttiin huomattavasti monimutkaisemmasta muistaakseni mm. perffisyistä: monimutkaisempi regex jumitti selaimen aivan täysin kun validoitavan emailin pituus oli kenties jossain 30-40 merkin tienoilla. Vahvat eksponentiaalisen kasvun vibat tuli siitä.
> 
> Sitten tietysti on myös se ongelma, että validaatiolla ei voi varmistua siitä onko mailiosoite oikeasti käytössä / kuuluu kyseiselle henkilölle. Ei toki tarkoita etteikö nykytilaa voisi parantaa. :)

Joo, toi linkin js regex vaikutti toimivan ihan ok vielä 100 merkkiin asti  ja 6x cpu throttle. Mut tota voi miettiä omassa tiksussaki tarvittaessa

pkoivisto commented:
> Joo, toi linkin js regex vaikutti toimivan ihan ok vielä 100 merkkiin asti ja 6x cpu throttle. Mut tota voi miettiä omassa tiksussaki tarvittaessa

Joop. Jatketaan tästä tosiaan vaikka toisen tiketin yhteydessä. Tuntuisi vähän väärältä sneakata tässä yhteydessä siihen muutoksia, kun tätä on tosiaan jo aiemminkin käsitelty muutamaan otteeseen. :)

pkoivisto committed:
Review feedback: handle undefined string within Utils.inspectCustomTextFieldErrors(..)

kimmotaskinen reviewed (approved):
Jees, näyttäs hyvältä!
